### PR TITLE
[WIP] IEEE 802.1AS gPTP support

### PR DESCRIPTION
**Pull request by jukkar**
This PR implements gPTP (generic Precision Time Protocol) support found in IEEE 802.1AS. This protocol can be used for time synchronization in ethernet networks. This PR is still very much work-in-progress and 
it builds on top of other not-yet merged code:

* PR #6601 Checksum offloading
* PR #6076 Traffic class support
* PR #5688 VLAN support

There are lot of patches in this PR, but if reviewing the code, the patch named "net: if: Add TX timestamp callback support" is the first one in this gPTP series.

The gPTP support has been developed on FRDM-K64F which has PTP and hw time-stamp supported in the HAL driver.

### CODE REVIEW

File: doc/subsystems/networking/gptp_release_notes.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+a 802.1AS bridge. However, this mode of operation has not been validated.
+
+Supported hardware
+******************
+
+Although the stack itself is hardware independent, ethernet frame timestamping
+support must be enabled in ethernet drivers.
+
+Boards supported:
+- NXP FRDM-K64F
```
Comment (dbkinder) [inline doc/subsystems/networking/gptp_release_notes.rst]: need a blank line before a list

File: doc/subsystems/networking/gptp_release_notes.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+- Qemu (emulated, limited capabilities due to lack of hardware clock)
```
Comment (dbkinder) [inline doc/subsystems/networking/gptp_release_notes.rst]: s/Qemu/QEMU/

File: doc/subsystems/networking/gptp_release_notes.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+Enabling the stack
+******************
+
+In menuconfig, following configuration must me enabled:
```
Comment (dbkinder) [inline doc/subsystems/networking/gptp_release_notes.rst]: ``the following configuration must be``

File: doc/subsystems/networking/gptp_release_notes.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+-CONFIG_NET_GPTP (Networing -> Link layer and IP networking support -> IP stack -> Link layer options -> Enable IEEE802.1AS support)
+
+Application interfaces
+**********************
+
+Only two Application Interfaces as defined in section 9 of the standard
+are available:
+-ClockTargetPhaseDiscontinuity interface (gptp_event_capture)
```
Comment (dbkinder) [inline doc/subsystems/networking/gptp_release_notes.rst]: add a blank line before this first list item

File: include/net/gptp.h [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+/**
+ * @file
+ * @brief Public functions for the Precision Time Protocol Stack.
+ *
+ */
+
+#ifndef __GPTP_H
+#define __GPTP_H
+
```
Comment (dbkinder) [inline include/net/gptp.h]: Need to define a doxygen defgroup to have this API show up in the documentation (and need a reference added to one of the .rst files in doc/api)

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
Added gptp support to native-posix board

**Comment by jukkar** [comment]
Rebase against master

**Comment by jukkar** [comment]
Lets remove this for time being as this seems to confuse people and is too early.

**Pull request closed without merging**